signer/src/stacks/api.rs

```diff
@@ -633,76 +625,120 @@ impl StacksClient {
-        )
-        .await;
-
-        let checked_result = check_result(self, result)?;
-
-        match checked_result {
-            Value::Sequence(SequenceData::List(ListData { data, .. })) => {
-                let mut keys = Vec::new();
-                for item in data {
-                    if let Value::Sequence(SequenceData::Buffer(BuffData { data })) = item {
-                        keys.push(PublicKey::from_slice(&data)?);
-                    } else {
-                        return Err(Error::InvalidStacksResponse(
-                            "expected a buffer but got something else",
-                        ));
+        self.exec(|client, retry| async move {
+            // Make a request to the sbtc-registry contract to get the current
+            // signer set.
+            let result = StacksClient::get_data_var(
+                client,
+                contract_principal,
+                &ContractName::from("sbtc-registry"),
+                &ClarityName::from("current-signer-set"),
+            )
+            .await?;
+
+            match result {
+                Value::Sequence(SequenceData::List(ListData { data, .. })) => {
+                    let mut keys = Vec::new();
+                    for item in data {
+                        if let Value::Sequence(SequenceData::Buffer(BuffData { data })) = item {
+                            keys.push(PublicKey::from_slice(&data)?);
+                        } else {
+                            retry.abort();
+                            return Err(Error::InvalidStacksResponse(
+                                "expected a buffer but got something else",
+                            ));
+                        }
                     }
+                    Ok(keys)
+                }
+                _ => {
+                    retry.abort();
+                    Err(Error::InvalidStacksResponse(
+                        "expected a sequence but got something else",
+                    ))
                 }
-                Ok(keys)
             }
-            _ => Err(Error::InvalidStacksResponse(
-                "expected a sequence but got something else",
-            )),
-        }
+        })
+        .await
     }
-    async fn get_account(&mut self, address: &StacksAddress) -> Result<AccountInfo, Error> {
-        check_result(self, StacksClient::get_account(self, address).await)
+
+    async fn get_account(&self, address: &StacksAddress) -> Result<AccountInfo, Error> {
+        self.exec(|client, _| async { StacksClient::get_account(client, address).await })
+            .await
     }
-    async fn submit_tx(&mut self, tx: &StacksTransaction) -> Result<SubmitTxResponse, Error> {
-        check_result(self, StacksClient::submit_tx(self, tx).await)
+
+    async fn submit_tx(&self, tx: &StacksTransaction) -> Result<SubmitTxResponse, Error> {
+        self.exec(|client, _| async { StacksClient::submit_tx(client, tx).await })
+            .await
```

Comment:
Yeah, before it was only on connection errors where the client would fail over. I felt that there's probably some more cases where we want to fail over, i.e. a 500 error or another type of error indicating corrupt chainstate etc.

My thought is that if the transaction truly can't be accepted (which really shouldn't happen, at least in the 99%), then even if it fails over and retries on the next node(s) then it will still fail. Since I haven't gone and dug into what kinds of errors are retry-able vs. node-specific vs. not retry-able, I removed those checks for now.

Do we want a todo to revisit this? Or should I have it check specifically for connection errors?

Reply:
Na, no need.